### PR TITLE
Add basic support for recognizing and validating Network Identity certs

### PR DESCRIPTION
**Pull request by ksperling-apple**
This involves changing some Extract*FromCert methods from using ChipCertificateSet to load the certificate to just using DecodeChipCert, since ChipCertificateSet makes additional assumptions.

Also remove unused ChipCertificateData.mCertificate member and make BitFlags more constexpr friendly.

Part of #29856

### CODE REVIEW

File: src/credentials/CHIPCert.cpp [hunk 1]
```diff
@@ -1446,5 +1427,33 @@ CHIP_ERROR CertificateValidityPolicy::ApplyDefaultPolicy(const ChipCertificateDa
     }
 }
 
+CHIP_ERROR ValidateChipNI(const ByteSpan & ni)
```
Comment (tcarmelveilleux) [inline src/credentials/CHIPCert.cpp:1430]: Please let's not use "Chip" instead of "Matter" for any new methods.

File: src/credentials/CHIPCert.h [hunk 2]
```diff
@@ -529,6 +530,14 @@ CHIP_ERROR VerifyCertSignature(const ChipCertificateData & cert, const ChipCerti
  */
 CHIP_ERROR ValidateChipRCAC(const ByteSpan & rcac);
 
+/**
+ * Validates a Network (Client) Identity in TLV-encoded form.
+ *
+ * This function parses the certificate, ensures the rigid fields have the values mandated by the
+ * specification, and validates the certificate signature.
+ */
+CHIP_ERROR ValidateChipNI(const ByteSpan & ni);
```
Comment (tcarmelveilleux) [inline src/credentials/CHIPCert.h:539]: - Please rename `ni` to `network_identity_cert` or a similarly readable name.
- Please document the error for invalid format, and for failed signature, and for success of validation.

File: src/credentials/CHIPCert_Internal.h [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+#include <credentials/CHIPCert.h>
+
+namespace chip {
+namespace Credentials {
+
+inline constexpr CharSpan kNetworkIdentityCN           = "*"_span;
+inline constexpr ByteSpan kNetworkIdentitySerialNumber = ByteSpan((uint8_t[1]){ 1 });
+
+inline constexpr uint32_t kNetworkIdentityNotBeforeTime = 1;
+inline constexpr uint32_t kNetworkIdentityNotAfterTime  = kNullCertTime;
```
Comment (tcarmelveilleux) [inline src/credentials/CHIPCert_Internal.h:30]: Do you have a spec section where the formatting is described to help cross-check?

File: src/credentials/CHIPCert_Internal.h [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+#pragma once
+
+#include <credentials/CHIPCert.h>
+
+namespace chip {
+namespace Credentials {
+
+inline constexpr CharSpan kNetworkIdentityCN           = "*"_span;
+inline constexpr ByteSpan kNetworkIdentitySerialNumber = ByteSpan((uint8_t[1]){ 1 });
```
Comment (tcarmelveilleux) [inline src/credentials/CHIPCert_Internal.h:27]: Requiring value `1` prevents ever allowing revocation. We should never force the serial number of the cert to be == 1. The CA should be able to set the serial number as needed.

File: src/credentials/CHIPCert_Internal.h [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+ */
+
+#pragma once
+
+#include <credentials/CHIPCert.h>
+
+namespace chip {
+namespace Credentials {
+
+inline constexpr CharSpan kNetworkIdentityCN           = "*"_span;
```
Comment (tcarmelveilleux) [inline src/credentials/CHIPCert_Internal.h:26]: Why this fixed CN? There are other ways to identify that a cert is a network identity CN, and that makes debugging easier.

File: src/credentials/CHIPCert.h [hunk 6]
```diff
@@ -529,6 +530,14 @@ CHIP_ERROR VerifyCertSignature(const ChipCertificateData & cert, const ChipCerti
  */
 CHIP_ERROR ValidateChipRCAC(const ByteSpan & rcac);
 
+/**
+ * Validates a Network (Client) Identity in TLV-encoded form.
+ *
+ * This function parses the certificate, ensures the rigid fields have the values mandated by the
+ * specification, and validates the certificate signature.
```
Comment (tcarmelveilleux) [inline src/credentials/CHIPCert.h:537]: Please provide at least a section name (number is too unstable over time).

Please explain how we trust the public key. We can't just magically trust the public key, given the cert has a "garbage" issuer field. Issuer == Subject always means that it's a self-signed cert, which implies a root cert and the key is not super relevant. This is very unorthodox without a lot of explanation.

**Review by tcarmelveilleux — CHANGES_REQUESTED** [review]
- [x] Please better document the non-standard public key crypto used here...

**Review by woody-apple — APPROVED** [review]
Approving pending comments from @tcarmelveilleux

File: src/credentials/CHIPCertToX509.cpp [hunk 7]
```diff
@@ -511,11 +511,12 @@ static CHIP_ERROR DecodeConvertTBSCert(TLVReader & reader, ASN1Writer & writer,
...
 
-        ReturnErrorOnFailure(reader.Next(kTLVType_ByteString, ContextTag(kTag_SerialNumber)));
-
         // serialNumber CertificateSerialNumber
         // CertificateSerialNumber ::= INTEGER
-        ReturnErrorOnFailure(writer.PutValue(kASN1TagClass_Universal, kASN1UniversalTag_Integer, false, reader));
+        ReturnErrorOnFailure(reader.Next(kTLVType_ByteString, ContextTag(kTag_SerialNumber)));
+        ReturnErrorOnFailure(reader.Get(certData.mSerialNumber));
+        ReturnErrorOnFailure(writer.PutValue(kASN1TagClass_Universal, kASN1UniversalTag_Integer, false,
+                                             certData.mSerialNumber.data(), static_cast<uint16_t>(certData.mSerialNumber.size())));
```
[the remark this answers is not in the document]
Comment (ksperling-apple) [reply src/credentials/CHIPCertToX509.cpp:519]: Yes this is just retaining the serial number ByteSpan in addition to passing it through to the ASN.1

File: src/credentials/CHIPCert.cpp (hunk 1 above)
Comment (woody-apple) [reply src/credentials/CHIPCert.cpp:1430]: My personal vote is keep it consistent. If in future we move chip to matter, then we do that once, and rip the bandaid off. For now, everything is chip, and I suggest keeping it that way.

Comment (ksperling-apple) [reply src/credentials/CHIPCert.cpp:1430]: I will call this `ValidateNetworkIdentity` for now, probably doesn't need either CHIP or Matter in the name.

Comment (ksperling-apple) [reply src/credentials/CHIPCert.cpp:1430]: On second thoughts, the `Chip` in the name here and other methods like `ValidateChipRCAC`, `ConvertChipCertToX509Cert` refers to the data format (Chip TLV vs X.509 DER), so I should probably keep it.

File: src/credentials/CHIPCertToX509.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (ksperling-apple) [reply src/credentials/CHIPCertToX509.cpp:519]: In terms of testing, this code is exercised by a lot of test cases in `TestChipCert`, so if I had broken the forwarding of the serial number into the ASN.1 output that would show up immediately. mSerialNumber being correctly populated is relied upon and tested via the Network Identity validation function.

File: src/credentials/CHIPCert_Internal.h (hunk 4 above)
Comment (ksperling-apple) [reply src/credentials/CHIPCert_Internal.h:27]: Discussed offline. Added a reference to the "Wi-F Authentication with Per-Device Credentials" spec section to cover the "why" questions.

File: src/credentials/CHIPCert_Internal.h (hunk 5 above)
Comment (ksperling-apple) [reply src/credentials/CHIPCert_Internal.h:26]: Discussed offline. Added a reference to the "Wi-F Authentication with Per-Device Credentials" spec section to cover the "why" questions.

File: src/credentials/CHIPCert.h (hunk 6 above)
Comment (ksperling-apple) [reply src/credentials/CHIPCert.h:537]: Discussed offline. Added a reference to the "Wi-F Authentication with Per-Device Credentials" spec section to cover the "why" questions.

**Review by tcarmelveilleux — APPROVED** [review]

**Pull request merged**